### PR TITLE
ast_match: shared AST-match helpers (PR 1a foundation, additive)

### PR DESCRIPTION
**Pull request by borisbat**
## Summary

PR 1a of the multi-PR linq_fold dedup plan — **foundation slice**. Adds shared AST-match helpers to `daslib/ast_match.das` + adds `push_block_list` to `daslib/templates_boost.das` next to `qmacro`. **Pure additive** — no consumer (linq_fold, sqlite_linq, decs_boost) is touched in this PR. Helpers were harvested from `daslib/linq_fold.das` + `modules/dasSQLITE/daslib/sqlite_linq.das`, where they appeared as private duplicates.

Migration of the consumers to these helpers is PR 1b (sqlite_linq) and PR 1c (linq_fold). Decs_boost migration is PR 6.

### `daslib/ast_match.das` (public)

| Helper | Promoted from | Purpose |
|---|---|---|
| `match_call_in_module(expr, name, modName) : ExprCall?` | `linq_fold:37 is_call_or_generic` + `sqlite_linq:580 match_module_call` | Match an `ExprCall` resolved to `(name, modName)`, transparent to generic instantiation via `func.fromGeneric` fallback. |
| `match_call_in_linq(expr, name) : ExprCall?` | `sqlite_linq:567 match_linq_call` | Wrapper on H1 with `"linq"`. |
| `peel_lambda_single_return(lam) : Expression?` | `sqlite_linq:557 peel_lambda_body` | Body of a single-arg, single-return lambda. |
| `peel_lambda_rename_var(expr, argName) : Expression?` | `linq_fold:52 fold_linq_cond` | Peel single-arg lambda with rename; falls back to `invoke(...)`. |
| `peel_lambda_replace_var(expr, replacement) : Expression?` | `linq_fold:71 fold_linq_cond_peel` | Peel with peeling-aware substitution; falls back to `invoke(...)`. |
| `peel_lambda_rename_2vars(expr, a, b) : Expression?` | `linq_fold:91 fold_linq_cond2` | 2-arg variant; returns `null` on non-peelable. |
| `peel_tuple_field_read(expr, bindName, fieldIndex) : bool` | `linq_fold:2095 is_bind_field_access` | Recognize `<bindName>._<fieldIndex>` shape. Drops the over-defensive `while`-peel for `ExprRef2Value` (never nests in the AST). |
| `qn(prefix, at) : string` | NEW | Collapses the 54-site `` "`{prefix}`{at.line}`{at.column}" `` qname construction. |

### `daslib/templates_boost.das` (public, next to `qmacro`)

| Helper | Purpose |
|---|---|
| `push_block_list(stmts, blockExpr)` | Splice every statement of a `qmacro_block(...)` result into `stmts`, cloning each. Replaces clusters of per-statement `stmts \|> push <\| qmacro_expr() { single }` calls (target: ~45 cluster sites in linq_fold). |

## Tests (9 new files, all green)

- `tests/ast_match/test_qn.das` — pure: basic, prefix/position uniqueness, LineInfo-zero edge case (documented: produces `` `prefix`0x0`0x0 ``, collides across synth sites with same prefix).
- `tests/ast_match/test_peel_lambda_single_return.das` — direct construction via `qmacro`.
- `tests/ast_match/test_peel_lambda_rename_var.das` — incl. rename keeps outer captures untouched, `ExprInvoke` fallback shape, clone discipline (input not mutated).
- `tests/ast_match/test_peel_lambda_replace_var.das` — needs typer-inserted `ExprRef2Value` wrapper; test manually wraps the buried `x` in `x + 1` since `apply_template` mutates children in place but cannot replace the cloned root (matches production usage).
- `tests/ast_match/test_peel_lambda_rename_2vars.das` — null-on-non-peelable contract.
- `tests/ast_match/test_peel_tuple_field_read.das` — direct `ExprField` / `ExprVar` / `ExprRef2Value` construction matrix.
- `tests/ast_match/test_match_call_in_module.das` + `test_match_call_in_linq.das` — structural / null-guard only (positive matches need typed AST, see scope note in test files).
- `tests/template/test_push_block_list.das` — multi-statement append, interleaved pushes, pipe form, null no-op.

CMakeLists glob auto-discovers — no wiring change required.

## Scope deviations from the plan

- **H7 (`extract_const_string`) deferred to PR 1b.** sqlite_linq has a byte-identical private version; adding the public duplicate creates ambiguity at sqlite_linq's call sites. PR 1b will promote it together with deleting the local copy.
- **Pre-existing lint cleanup**: fixed 2 PERF020 (redundant `string()` cast on `das_string`) in `daslib/ast_match.das` per the touched-file lint discipline.

## Test plan

- [x] `mcp__daslang__compile_check` all touched files PASS
- [x] `mcp__daslang__format_file` all touched files already-formatted
- [x] `mcp__daslang__lint` all touched files clean (11 files, 0 issues)
- [x] `mcp__daslang__run_test` for each of the 9 new tests PASS
- [x] `mcp__daslang__compile_check tests/ast_match/*.das` — all 28 existing+new files compile (no regression to existing qmatch suite)
- [x] `mcp__daslang__compile_check daslib/linq_fold.das, modules/dasSQLITE/daslib/sqlite_linq.das` — both downstream consumers compile unchanged
- [ ] Full CI matrix (interp + AOT + JIT, all platforms)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: daslib/ast_match.das [hunk 1]
```diff
@@ -2191,3 +2191,142 @@ class private QMatchFunction : AstCallMacro {
...
+    if (expr == null) return false
+    var sub = expr
+    if (sub is ExprRef2Value) {
+        sub = (sub as ExprRef2Value).subexpr
+    }
+    if (!(sub is ExprField)) return false
+    let fld = sub as ExprField
+    if (fld.name != "_{fieldIndex}") return false
+    var src = fld.value
+    if (src is ExprRef2Value) {
```
[the remark this answers is not in the document]
Comment (borisbat) [reply daslib/ast_match.das:2307]: Pushing back — the direction here is intentional, and the codebase convention is going the other way:

`ExprRef2Value` never nests in the AST. The typer inserts a single wrapper around a reference-typed read; that wrapper is never wrapped again by any pass (macros / templates / linters). So `while (... is ExprRef2Value)` is over-defended — a single `if`-peel is sufficient and correct.

The mentioned `qm_peel_ref2value` and the 4 `while`-loop sites in `daslib/linq_fold.das` (2100, 2108, 2510, 2548) are the pre-existing inconsistency, not the new helper. Adding the cleanup to PR 2 of this multi-PR plan: simplify `qm_peel_ref2value` to a single `if` and either drop the linq_fold while-loops entirely (via the new `peel_tuple_field_read` helper + qmatch's auto-peel) or route them through `qm_peel_ref2value` as a single source of truth. The new helper's comment already pins the invariant; I'll mirror it on `qm_peel_ref2value` when that PR lands.

Wontfix in this PR.
